File: praktikum/praktikum-1/3-Kompleks/Kompleks.cpp
/* 	NIM			: 13521129
   	Nama     	: Chiquita Ahsanunnisa
   	Tanggal		: 9 Februari 2023
   	Topik		: Praktikum 1
   	Deskripsi	: Kompleks */

#include "Kompleks.h"
#include <iostream>
using namespace std;

int Kompleks::n_kompleks = 0;

// ctor tanpa parameter
// inisialisasi seluruh koefisien dengan nilai 0
Kompleks::Kompleks() {
	this->real = 0;
	this->imaginer = 0;
	n_kompleks++;
}

// ctor dengan parameter
Kompleks::Kompleks(int real, int imaginer) {
	this->real = real;
	this->imaginer = imaginer;
	n_kompleks++;
}

// mengembalikan bagian riil
int Kompleks::GetReal() const {
	return this->real;
}

// mengembalikan bagian imaginer
int Kompleks::GetImaginer() const {
	return this->imaginer;
}

// mengisi bagian riil
void Kompleks::SetReal(int val) {
	this->real = val;
}

// mengisi bagian imaginer
void Kompleks::SetImaginer(int val) {
	this->imaginer = val;
}

// operator overloading

// operator+ untuk melakukan penjumlahan dengan rumus berikut
// (a + bi) + (c + di) = (a+c) + (b+d)i
Kompleks operator+ (const Kompleks& k1, const Kompleks& k2) {
	return Kompleks(k1.real + k2.real, k1.imaginer + k2.imaginer);
}

// operator- untuk melakukan pengurangan dengan rumus berikut
// (a + bi) − (c + di) = (a−c) + (b−d)i
Kompleks operator- (const Kompleks& k1, const Kompleks& k2) {
	return Kompleks(k1.real - k2.real, k1.imaginer - k2.imaginer);
}

// operator* untuk melakukan perkalian dengan rumus berikut
// (a + bi)(c + di) = ac + bci + adi + bd i^2 = (ac−bd) + (bc+ad)i
Kompleks operator* (const Kompleks& k1, const Kompleks& k2) {
	return Kompleks(k1.real * k2.real - k1.imaginer * k2.imaginer, k1.imaginer * k2.real + k1.real * k2.imaginer);
}

// operator* untuk mengkalikan bilangan kompleks dengan konstanta
// (a + bi)(c) = (ac) + (bc)i
Kompleks operator* (const Kompleks& kom, const int c) {
	return Kompleks(c * kom.real, c * kom.imaginer);
}

// operator* untuk mengkalikan bilangan kompleks dengan konstanta (sifat komutatif)
Kompleks operator* (const int c, const Kompleks& kom) {
	return Kompleks(c * kom.real, c * kom.imaginer);
}

// mengembalikan jumlah instance yang pernah dibuat
int Kompleks::CountKompleksInstance() {
	return n_kompleks;
}

// mencetak bilangan kompleks ke layar, diakhiri dengan end-of-line 
// contoh:
// 3+5i
// 0+0i
// -5-4i
void Kompleks::Print() {
	cout << this->real << (this->imaginer < 0 ? '-' : '+') << (this->imaginer < 0 ? -1 * this->imaginer : this->imaginer) << 'i' << endl;
}

File: responsi/responsi-4/1-Kompleks/Kompleks.java
/* 	NIM			: 13521129
   	Nama     	: Chiquita Ahsanunnisa
   	Tanggal		: 16 Maret 2023
   	Topik		: Responsi 4
   	Deskripsi	: Kompleks */

class Kompleks {
    private static int n_kompleks;
    private int real, imaginer;

    // ctor tanpa parameter
    // inisialisasi seluruh koefisien dengan nilai 0
    public Kompleks() {
        real = 0;
        imaginer = 0;
        n_kompleks++;
    }

    // ctor dengan parameter
    public Kompleks(int real, int imaginer) {
        this.real = real;
        this.imaginer = imaginer;
        n_kompleks++;
    }

    //mengembalikan bagian riil
    public int getReal() { return real; }

    // mengembalikan bagian imaginer
    public int getImaginer() { return imaginer; }

    // mengisi bagian riil
    public void setReal(int real) { this.real = real; }

    // mengisi bagian imaginer
    public void setImaginer(int imaginer) { this.imaginer = imaginer; }

    // operator+ untuk melakukan penjumlahan dengan rumus berikut
    // (a + bi) + (c + di) = (a+c) + (b+d)i
    public Kompleks plus(Kompleks b) {
        return new Kompleks(this.real + b.real, this.imaginer + b.imaginer);
    }

    // operator- untuk melakukan pengurangan dengan rumus berikut
    // (a + bi) − (c + di) = (a−c) + (b−d)i
    public Kompleks minus(Kompleks b) {
        return new Kompleks(this.real - b.real, this.imaginer - b.imaginer);
    }

    // operator* untuk melakukan perkalian dengan rumus berikut
    // (a + bi)(c + di) = ac + bci + adi + bd i^2 = (ac−bd) + (bc+ad)i
    public Kompleks multiply(Kompleks b) {
        return new Kompleks(this.real * b.real - this.imaginer * b.imaginer, this.imaginer * b.real + this.real * b.imaginer);
    }

    // operator* untuk mengkalikan bilangan kompleks dengan konstanta
    // (a + bi)(c) = (ac) + (bc)i
    public Kompleks multiply (int c) {
        return new Kompleks(this.real * c, this.imaginer * c);
    }

    // mencetak bilangan kompleks ke layar, diakhiri dengan end-of-line
      // contoh:
      // 3+5i
      // 0
      // 3i
      // -3
      // -5-4i
    public void print() {
        System.out.print(this.real);
        if (this.imaginer > 0) {
            System.out.print("+" + this.imaginer + "i");
        } else if (this.imaginer < 0) {
            System.out.print("-" + (-1) * this.imaginer + "i");
        }
        System.out.print("\n");
    }

    public static int countKompleksInstance() { return n_kompleks; }
}
